source/common/common/shared_memory_hash_set.h

```diff
@@ -0,0 +1,370 @@
+   * @param key the key to remove
+   */
+  bool remove(absl::string_view key) {
+    const uint32_t slot = computeSlot(key);
+    uint32_t* next = nullptr;
+    for (uint32_t* cptr = &slots_[slot]; *cptr != Sentinal; cptr = next) {
+      Cell& cell = getCell(*cptr);
+      next = &cell.next_cell;
+      if (cell.value.key() == key) {
+        control_->free_cell_index = *cptr;
+        --control_->size;
+        *cptr = *next; // Splices current cell out of slot-chain.
+        *next = control_->free_cell_index;
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /** Returns the number of key/values stored in the map. */
+  uint32_t size() const { return control_->size; }
+
+  /**
+   * Gets the value associated with a key, returning nullptr if the value was not found.
+   * @param key
+   */
+  Value* get(absl::string_view key) {
+    const uint32_t slot = computeSlot(key);
+    for (uint32_t c = slots_[slot]; c != Sentinal; c = getCell(c).next_cell) {
+      Cell& cell = getCell(c);
+      if (cell.value.key() == key) {
+        return &cell.value;
+      }
+    }
+    return nullptr;
+  }
+
+private:
+  uint32_t computeSlot(absl::string_view key) {
+    return HashUtil::xxHash64(key) % options_.num_slots;
+  }
+
+  /**
+   * Computes a signature string, composed of all the non-zero 8-bit characters.
+   * This is used for detecting if the hash algorithm changes, which invalidates
+   * any saved stats-set.
+   */
+  std::string signatureStringToHash() {
+    std::string signature_string;
+    signature_string.resize(255);
+    for (int i = 1; i <= 255; ++i) {
+      signature_string[i - 1] = i;
+    }
+    return signature_string;
+  }
+
+  /**
+   * Represents control-values for the hash-table, including a mutex, which
+   * must gate all access to the internals.
+   */
+  struct Control {
+    std::string toString() const {
+      return fmt::format("{} size={} free_cell_index={}", options.toString(), size,
+                         free_cell_index);
+    }
+
+    SharedMemoryHashSetOptions options; // Options established at map construction time.
+    uint64_t hash_signature;            // Hash of a constant signature string.
+    uint32_t num_bytes;                 // Bytes allocated on behalf of the map.
+    uint32_t size;                      // Number of values currently stored.
+    uint32_t free_cell_index;           // Offset of first free cell.
+  };
+
+  /**
+   * Represents a value-cell, which is stored in a linked-list from each slot.
+   */
+  struct Cell {
+    uint32_t next_cell; // Offset of next cell in map->cells_, terminated with Sentinal.
+    Value value;        // Templated value field.
+  };
+
+  // It seems like this is an obvious constexpr, but it won't compile as one.
+  static size_t alignment() {
+    return std::max(alignof(Cell), std::max(alignof(uint32_t), alignof(Control)));
+  }
+
+  static uint32_t align(uint32_t size) { return (size + alignment() - 1) & ~(alignment() - 1); }
```

Comment:
Please RELEASE_ASSERT that alignment() returns a power-of-2; otherwise this math doesn't work (I think)